components/sup/src/manager/file_watcher.rs

```diff
@@ -340,25 +341,31 @@ impl<C: Callbacks, W: Watcher> FileWatcher<C, W> {
             // this should only ever trigger the one time
         }
 
-        // If the WatchedFileState is NonExistent when this method begins then there are no watches
-        // with which the notify crate send us events so the WatchedFileState must be manually
-        // assessed and if the new state is ExistentFile then file appearance needs to be signaled.
-        // If the new state is ExistendDirectory then on the next iteration there will be a watch
-        // in place.  In any event when the WatchedFileState is NonExistent this method can short
-        // circuit and exit without trying to listen.
-        if self.watched_file.state == WatchedFileState::NonExistent {
-            self.manage_watch()?;
-            if self.watched_file.state == WatchedFileState::ExistentFile {
-                debug!("-=> WatchedFileState::ExistentFile");
+        let current_state = &self.watched_file.state;
+        let next_state = self.watched_file.assess_state();
+        trace!("| ({:?},{:?})", current_state, next_state);
+
+        match (current_state, &next_state) {
+            (WatchedFileState::ExistentFile, WatchedFileState::ExistentFile)
+            | (WatchedFileState::ExistentFile, WatchedFileState::ExistentDirectory)
+            | (WatchedFileState::ExistentFile, WatchedFileState::NonExistent)
+            | (WatchedFileState::ExistentDirectory, WatchedFileState::ExistentDirectory)
+            | (WatchedFileState::ExistentDirectory, WatchedFileState::ExistentFile)
+            | (WatchedFileState::ExistentDirectory, WatchedFileState::NonExistent) => {
+                match self.rx.try_recv() {
+                    Ok(notify_result) => self.handle_event(&notify_result?),
+                    Err(TryRecvError::Empty) => Ok(()),
+                    Err(e) => Err(e.into()),
+                }
+            }
+            (WatchedFileState::NonExistent, WatchedFileState::ExistentFile) => {
                 self.callbacks.file_appeared(&self.watched_file.path);
+                self.manage_watch_with_next_state_as(next_state)
             }
-            return Ok(());
-        }
-
-        match self.rx.try_recv() {
-            Ok(notify_result) => self.handle_event(&notify_result?),
-            Err(TryRecvError::Empty) => Ok(()),
-            Err(e) => Err(e.into()),
+            (WatchedFileState::NonExistent, WatchedFileState::ExistentDirectory) => {
+                self.manage_watch_with_next_state_as(next_state)
+            }
+            (WatchedFileState::NonExistent, WatchedFileState::NonExistent) => Ok(()),
```

Comment:
seems like this could be simplified with something like:
```
        match current_state {
            WatchedFileState::NonExistent => {
                match &next_state {
                    WatchedFileState::ExistentFile => {
                        self.callbacks.file_appeared(&self.watched_file.path);
                        self.manage_watch_with_next_state_as(next_state)
                    }
                    WatchedFileState::ExistentDirectory => {
                        self.manage_watch_with_next_state_as(next_state)
                    }
                    WatchedFileState::NonExistent => Ok(()), 
                }   
            }
            _ => {
                match self.rx.try_recv() {
                    Ok(notify_result) => self.handle_event(&notify_result?),
                    Err(TryRecvError::Empty) => Ok(()),
                    Err(e) => Err(e.into()),
                }
            }
        }
```

Reply:
Here I really like the explicitness of the (current_state, next_state) pairs. And for some reason I preferred to keep the cases explicitly enumerated but I can't justify why, especially given that I wouldn't always break that way. So based on your feedback I went with the following alternative.
```
        match (current_state, &next_state) {
            (WatchedFileState::NonExistent, WatchedFileState::ExistentFile) => {
                self.callbacks.file_appeared(&self.watched_file.path);
                self.manage_watch_with_next_state_as(next_state)
            }
            (WatchedFileState::NonExistent, WatchedFileState::ExistentDirectory) => {
                self.manage_watch_with_next_state_as(next_state)
            }
            (WatchedFileState::NonExistent, WatchedFileState::NonExistent) => Ok(()),
            (_, _) => {
                match self.rx.try_recv() {
                    Ok(notify_result) => self.handle_event(&notify_result?),
                    Err(TryRecvError::Empty) => Ok(()),
                    Err(e) => Err(e.into()),
                }
            }
        }
```
If that's fine just resolve this.  If not, let me know and I will use your suggested version.

Reply:
yes this works.